pennylane/gradients/hessian_transform.py

```diff
@@ -0,0 +1,235 @@
+                # a tuple of Jacobians, each with respect to one QNode argument, and each of shape:
+                #   (# gate args, qnode arg shape)
+                # The quantum Hessian qhess has shape:
+                #   (qnode output shape, # gate args, # gate args)
+                # Return cjac.T @ qhess @ cjac (in the scalar case).
+                hessians = []
+
+                for c in cjac:
+                    if c is not None:
+                        # number of dimensions in qnode arg shape
+                        num_arg_dims = len(qml.math.shape(c)) - 1
+                        # hess -> (qnode output shape, # gate args, qnode arg shape)
+                        hess = qml.math.tensordot(qhess, c, [[-1], [0]])
+                        # hess -> (qnode output shape, qnode arg shape, qnode arg shape)
+                        hess = qml.math.tensordot(hess, c, [[-1 - num_arg_dims], [0]])
+                        hessians.append(hess)
+
+                return tuple(hessians)
+
+            is_square = cjac.shape == (1,) or (cjac.ndim == 2 and cjac.shape[0] == cjac.shape[1])
+
+            if is_square and qml.math.allclose(cjac, qml.numpy.eye(cjac.shape[0])):
+                # Classical Jacobian is the identity. No classical processing
+                # is present inside the QNode.
+                return qhess
+
+            # Classical processing present of either:
+            #   a) a single argument or
+            #   b) multiple arguments of the same shape
+            # The shape of the classical jacobian returned by qml.jacobian (gets invoked by
+            # qml.transforms.classical_jacobian with autograd) depends on the scenario:
+            #   a) (# gate args, qnode arg shape)
+            #   b) (reverse qnode arg shape, # gate args, # qnode args)
+            # It then needs to be contracted twice (cjac.T @ hess @ cjac for scalar-valued QNodes)
+            # with the quantum Hessian of shape:
+            #      (qnode output shape, # gate args, # gate args)
+            # The result should have shape:
+            #   a) (qnode output shape, qnode arg shape, qnode arg shape)
+            #   b) (reverse qnode arg shape A, # qnode args B, qnode output shape, qnode arg shape B, # qnode args A)
+            num_gate_args = qml.math.shape(qhess)[-1]
+            # Consider only QNode arguments regarded as trainable by the interfaces.
+            trainable_args_idx = self._jacobian_trainable_args(args, qnode.interface)
+            if not trainable_args_idx:  # pragma: no cover
+                warnings.warn(
+                    "Trainable indices were not detected for the classical Jacobian, cannot "
+                    "compute the hybrid Hessian of the QNode. You can still attempt to obtain "
+                    "the quantum Hessian with the `hybrid=False` parameter.",
+                    UserWarning,
+                )
+                return ()
```

Comment:
I'm curious why an exception is not raised here, rather than a warning? The way I typically think about it, a warning is raised if the output _may_ not be correct just to warn the user?

Reply:
Ah, nevermind, I realize my mistake now - it _could_ be correct that the user passed a QNode with no trainable parameters :)

Just verifying that an empty tuple makes sense to return in this case? I have lost track of what we return when 😆

Reply:
Well if no trainable parameters are present we already return an empty tuple further up I think. This warning is similar to the [one below](https://github.com/dime10/pennylane/blob/1f7442d8fbb28c41187f7be213d98da5fd69157c/pennylane/gradients/hessian_transform.py#L225-L231), neither should in principle be invoked and they are there as a just in case (discussed in a [bugifx pr here](https://github.com/PennyLaneAI/pennylane/pull/1989#discussion_r766811031)). But we shouldn't worry about these as both will be removed with #2059.